File: examples/ruby/osgart.rb
#
# Demonstation for Ruby
#

# extend the load path for wosg
$LOAD_PATH << '../../bin/ruby'


# include all necessary libs
require 'wosg'
require 'wosgDB'
require 'wosgProducer'
require 'wosgART'

class Viewer

	def initialize()

    	# a producer
		@viewer = WosgProducer::Viewer.new

		# esc stops the viewer
		@viewer.setUpViewer(WosgProducer::Viewer::ESCAPE_SETS_DONE)

		# do not use near/far clipping
		@viewer.getCullSettings().setComputeNearFarMode(Wosg::CullSettings::DO_NOT_COMPUTE_NEAR_FAR)

		# configure the video
		cfg = WosgART::VideoConfiguration.new
		cfg.deviceconfig = "data/WDM_camera_flipv.xml"

		# create a video source
		@video = WosgART::VideoManager.createVideoFromPlugin("osgart_artoolkit",cfg)

		@video.open

		@tracker = WosgART::ARToolKitTracker.new

		WosgART::TrackerManager::getInstance().addTracker(@tracker)

		@tracker.init(@video.getWidth(),@video.getHeight())

		foreground = Wosg::Group.new
		foreground.getOrCreateStateSet().setRenderBinDetails(2, "RenderBin")

		vb = WosgART::VideoBackground.new(@video.getId())

		vb.init

		foreground.addChild(vb)

		projection = Wosg::Projection.new(Wosg::Matrixd.new(@tracker.getProjectionMatrix()))

		@marker_trans = WosgART::ARTTransform.new(0)

		@marker = @marker_trans.getMarker();

		@marker.setActive(true)

		@marker_trans.getOrCreateStateSet().setRenderBinDetails(5, "RenderBin")

		@marker_trans.addChild(WosgDB::readNodeFile("cow.osg"))

		modelview = Wosg::MatrixTransform.new

		modelview.addChild(foreground)

		foreground.addChild(@marker_trans)

		projection.addChild(modelview)

		# create a root node for the scene
		@root = Wosg::Group.new()

		@root.addChild(projection)

		@viewer.setSceneData(@root)

		@viewer.realize()

		@video.start

	end

	def run()

		while !@viewer.done()

			@viewer.sync

			@video.update

			@tracker.setImage(@video)
			@tracker.update

			@viewer.update

			@viewer.frame

		end

		@viewer.sync

		@viewer.cleanup_frame

		@video.stop

		@video.close


	end

end

v = Viewer.new
v.run

File: examples/lua/osgart.lua
--
-- example implementation for an OpenSceneGraph and osgART
--

-- extend the library search path
package.cpath = '../../bin/lua/?.so;../../bin/lua/?.dll;' .. package.cpath

require 'wosg'
require 'wosgProducer'
require 'wosgDB'
require 'wosgART'

viewer = {}

viewer.init = function()

	-- create instance of a viewer
	viewer.v = wosgProducer.Viewer()


	-- set default settings
	viewer.v:setUpViewer(wosgProducer.ESCAPE_SETS_DONE)

	-- AR does not need far and near clipping
   	viewer.v:getCullSettings():setComputeNearFarMode(wosg.DO_NOT_COMPUTE_NEAR_FAR)

	-- create a root node
	viewer.root = wosg.Group()

	-- osgART, open a video feed
	viewer.video = wosgART.Video("data/WDM_camera.xml")

	-- add the video stream, please note that static members
	-- are accessible through classname_membername in lua
	wosgART.VideoManager_getInstance():addVideoStream(viewer.video)

	-- open the video feed
	viewer.video:open()

	-- tracker instance
	viewer.art = wosgART.ARToolKitTracker()

	wosgART.TrackerManager_getInstance():addTracker(viewer.art)

	viewer.art:init(viewer.video:xSize(), viewer.video:ySize())


	-- background
	viewer.background = wosgART.VideoBackground(viewer.video:getId())

	viewer.background:init()

	-- add the background
	viewer.root:addChild(viewer.background)

	-- change render mode
	viewer.background:getOrCreateStateSet():setRenderBinDetails(0, "RenderBin",	osg.OVERRIDE_RENDERBIN_DETAILS)

	-- Magic
	-- viewer.root:getOrCreateStateSet():setRenderBinDetails(2, "RenderBin", wosg.OVERRIDE_RENDERBIN_DETAILS)

	-- get the transform
	viewer.markertrans = wosgART.ARTTransform(0)

	-- get the actual marker
	viewer.marker = viewer.markertrans:getMarker()

	-- make it visible
	viewer.marker:setActive(true)

	-- note: osg.Matrix is osg.Matrixd (we are using double precision)
	viewer.projection = wosg.Projection(wosg.Matrixd(viewer.art:getProjectionMatrix()))

	-- modelview
	viewer.modelscene = wosg.MatrixTransform()

	-- viewer.modelscene:getOrCreateStateSet():setRenderBinDetails(5, "RenderBin",	wosg.OVERRIDE_RENDERBIN_DETAILS)

	-- read file
	viewer.cow = wosgDB.readNodeFile("cow.osg")


	-- add to the marker
	-- viewer.markertrans:addChild(viewer.cow)
	viewer.modelscene:addChild(viewer.markertrans)
	viewer.projection:addChild(viewer.modelscene)

   	viewer.root:addChild(viewer.projection)


	-- add the root node to the viewer
	viewer.v:setSceneData(viewer.root)

end


viewer.run = function()

	-- run now
	io.write("--\nRun Function\n")

	-- create a viewer
	viewer.v:realize()

	viewer.video:start()

	-- run until the viewer manipulator does quit (ESC)
	while viewer.v:done() ~= true do

		-- sync the viewer
		viewer.v:sync()

		-- check for new video images
		viewer.video:update()

		-- set the image (maintain pointer)
    	viewer.art:setImage(viewer.video)
    	viewer.art:update()

    	if viewer.marker:isValid() then
			io.write("Cow!\n")
			-- local matrix = osg.Matrixd(viewer.marker:getTransform())
		else
			io.write("Grass!\n")
		end

		-- update the scenegraph
		viewer.v:update()

		-- call the renderer
		viewer.v:frame()
	end
end


viewer.exit = function()

	io.write("--\nExit Function\n")

	viewer.video:stop()
	viewer.video:close()
	viewer.v:sync()
	viewer.v:cleanup_frame();
end

-- run the example
viewer.init()
viewer.run()
viewer.exit()




